### ISSUE TITLE
FDB + SplinterDB engine: YCSB workloada (1.5 M rows, 12 threads) runs into fatal assertion: trunk_verify: pivot kv_bytes incorrect

### ISSUE THREAD

**Issue by gapisback** [vmware/splinterdb#424]
### Describe the bug

Execution of YCSB `workloada` with 1.5M recordcount (# of rows), executed with 12 Java threads runs into the following hard assertion during the run phase.

```
Jul 09 16:48:43 sdb-fdb-build fdbmonitor[637834]: LogGroup="default" Process="fdbserver.5500": trunk_verify: pivot kv_bytes incorrect
Jul 09 16:48:43 sdb-fdb-build fdbmonitor[637834]: LogGroup="default" Process="fdbserver.5500": reported 49070558, actual 50018047
```
Subsequently, after reboot we consistently run into the assertion reported under issue #305: 
```
Jul 09 16:51:40 : Assertion failed at src/clockcache.c:2138:clockcache_get_internal(): "(extent_ref_count > 1)".
```

Seems like the initial assertion in `trunk.c` is reporting some sort of a corruption which prevents a subsequent reboot from succeeding.

`load` phase goes through cleanly. 

We see this assertion [1] somewhere about 12 mins into the `run` phase. [2]

[1] Full dump when assertion trips is as follows:

```
Jul 09 16:37:43 sdb-fdb-build fdbmonitor[637834]: LogGroup="default" Process="fdbserver.5500": fingerprint_size: 27
Jul 09 16:37:43 : Fingerprint size 29 too large, max value size is 5, setting to 27
Jul 09 16:37:43 : filter-index-size: 256 is too small, setting to 512
Jul 09 16:48:43 : trunk_verify: pivot kv_bytes incorrect
Jul 09 16:48:43 : reported 49070558, actual 50018047
Jul 09 16:48:43 : addr: 1445888
Jul 09 16:48:43 : ----------------------------------------------------------------------------------------------------
Jul 09 16:48:43 : |          |     addr     |   next addr  | height |   gen   | pvt gen |                            |
Jul 09 16:48:43 : |  HEADER  |--------------|--------------|--------|---------|---------|----------------------------|
Jul 09 16:48:43 : |          |      1445888 |            0 |      1 |       0 |      10 |                            |
Jul 09 16:48:43 : |--------------------------------------------------------------------------------------------------|
Jul 09 16:48:43 : |                                       PIVOTS                                                     |
Jul 09 16:48:43 : |--------------------------------------------------------------------------------------------------|
Jul 09 16:48:43 : |         pivot key        |  child addr  |  filter addr | tuple count | kv bytes  |  srq  |  gen  |
Jul 09 16:48:43 : |--------------------------|--------------|--------------|-------------|-----------|-------|-------|
Jul 09 16:48:43 : | 0x0244423a75736572746162 |      1327104 |   7758413824 |       29669 |  35632269 |    -1 |     1 |
Jul 09 16:48:43 : | Full key: '0x0244423a757365727461626c653a757365723431303439313532333331393233323931333200'
Jul 09 16:48:43 : | 0x0244423a75736572746162 |      1314816 |   7758938112 |       41647 |  49070558 |    -1 |     7 |
Jul 09 16:48:43 : | Full key: '0x0244423a757365727461626c653a757365723531343236343033383938323032333830323200'
Jul 09 16:48:43 : | 0x0244423a75736572746162 |      1339392 |   7759462400 |       35500 |  40794367 |    -1 |     2 |
Jul 09 16:48:43 : | Full key: '0x0244423a757365727461626c653a757365723631383239353732383831343037383739353500'
Jul 09 16:48:43 : | 0x0244423a75736572746162 |      1318912 |            0 |           0 |         0 |    -1 |     9 |
Jul 09 16:48:43 : | Full key: '0x0244423a757365727461626c653a757365723732323137393131353536363733343137393400'
Jul 09 16:48:43 : | 0x0244423a75736572746162 |      1347584 |            0 |           0 |         0 |    -1 |     6 |
Jul 09 16:48:43 : | Full key: '0x0244423a757365727461626c653a757365723737343032353031313536323333313938363300'
Jul 09 16:48:43 : | 0x0244423a75736572746162 |      1335296 |   7760510976 |       44683 |  50956328 |    -1 |     0 |
Jul 09 16:48:43 : | Full key: '0x0244423a757365727461626c653a757365723832363131383239343038343030373032323900'
Jul 09 16:48:43 : | 0xffffffffffffffffffffff |              |              |             |           |       |       |
Jul 09 16:48:43 : | Full key: '0xfffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff'
Jul 09 16:48:43 : |--------------------------------------------------------------------------------------------------|
Jul 09 16:48:43 : |                              BRANCHES AND [SUB]BUNDLES                                           |
Jul 09 16:48:43 : |start_branch=49 end_branch=58 start_bundle=9  end_bundle=10 start_sb=9  end_sb=10                 |
Jul 09 16:48:43 : |--------------------------------------------------------------------------------------------------|
Jul 09 16:48:43 : |   # |          point addr         | filter1 addr | filter2 addr | filter3 addr |                 |
Jul 09 16:48:43 : |     |    pivot/bundle/subbundle   |  num tuples  |              |              |                 |
Jul 09 16:48:43 : |-----|--------------|--------------|--------------|--------------|--------------|-----------------|
Jul 09 16:48:43 : |     |        -- pivot  5 --       |              |              |              |                 |
Jul 09 16:48:43 : |  49 |           6858604544        |              |              |              |                 |
Jul 09 16:48:43 : |     |        -- pivot  1 --       |              |              |              |                 |
Jul 09 16:48:43 : |  50 |           7001604096        |              |              |              |                 |
Jul 09 16:48:43 : |  51 |           7092305920        |              |              |              |                 |
Jul 09 16:48:43 : |     |        -- pivot  2 --       |              |              |              |                 |
Jul 09 16:48:43 : |  52 |           7244087296        |              |              |              |                 |
Jul 09 16:48:43 : |  53 |           7334658048        |              |              |              |                 |
Jul 09 16:48:43 : |     |        -- pivot  0 --       |              |              |              |                 |
Jul 09 16:48:43 : |  54 |           7482114048        |              |              |
Jul 09 16:48:43 :      |                 |
Jul 09 16:48:43 : |  55 |           7629307904        |              |              |              |                 |
Jul 09 16:48:43 : |  56 |           7720009728        |              |              |              |                 |
Jul 09 16:48:43 : |     |       -- bundle  9 --       |       105809 |              |              |                 |
Jul 09 16:48:43 : |     |  --   comp subbundle  9 --  |   8530165760 |   8615231488 |            0 |                |
Jul 09 16:48:43 : |  57 |           8617459712        |              |              |              |                 |
Jul 09 16:48:43 : ----------------------------------------------------------------------------------------------------
Jul 09 16:48:43 :
Jul 09 16:48:43 : }
Jul 09 16:48:43 : Assertion failed at src/trunk.c:4940:trunk_compact_bundle(): "trunk_verify_node(spl, node)".
```

[2] Snippets of `run` outputs showing how far we progressed before the assertion tripped:

```
2022-07-09 16:47:54:353 0 sec: 0 operations; est completion in 0 second
DBWrapper: report latency for each error is false and specific error codes to track for latency are: []
2022-07-09 16:48:04:308 10 sec: 28769 operations; 2876.9 current ops/sec; est completion in 8 minutes [READ: Count=14395, Max=35743, Min=846, Avg=2629.76, 90=3639, 99=5283, 99.9=29567, 99.99=34655] [UPDATE: Count=14392, Max=264447, Min=2814, Avg=5600.93, 90=6863, 99=15047, 99.9=31391, 99.99=51583]
[...]
2022-07-09 17:00:14:308 740 sec: 154743 operations; 0 current ops/sec; est completion in 1 hour 47 minutes [READ: Count=0, Max=0, Min=9223372036854775807, Avg=NaN, 90=0, 99=0, 99.9=0, 99.99=0] [UPDATE: Count=0, Max=0, Min=9223372036854775807, Avg=NaN, 90=0, 99=0, 99.9=0, 99.99=0]
```

[3] Upon reboot, we consistently run into the assertion that was originally reported under issue #305 for which PR # 410 applied some instrumentation to track an assertion tripping in `clockcache_get_internal()`

```
Jul 09 16:51:40 : fingerprint_size: 27
Jul 09 16:51:40 : filter-index-size: 256 is too small, setting to 512
Jul 09 16:51:40 : FDBD joined cluster.
Jul 09 16:51:40 : Allocated 0 extents at mount: (0 MiB)
Jul 09 16:51:40 : Allocated extents: 0
Jul 09 16:51:40 :    Index  ExtentAddr  Count
Jul 09 16:51:40 : Found 0 extents with allocated pages.
Jul 09 16:51:40 : Assertion failed at src/clockcache.c:2138:clockcache_get_internal(): "(extent_ref_count > 1)".
Jul 09 16:51:40 : Attempt to get a buffer for page addr=4096, page type=5 ('superblock'), from extent addr=0, (extent number=0), which is an unallocated extent, extent_ref_count=0.
Jul 09 16:51:41 : Process 638468 terminated by signal 6, restarting in 56 seconds
```

### Reproduction steps

```bash
This will consistently reproduce if you run the Java-client YCSB bin/ycsb driver with FDB + SplinterdB integrated engine, as follows:

$ ./run-fdb-ycsb-tests -w workloada -r 1500000 -t 12 splinterdb

This script can be found on the `sdb-fdb-build` machine in the `agurajada/fdb-spdb-intgn-1` branch.

NOTE: This YCSB runs succeed cleanly for all workloads (a-f) with 1M # of rows.

At the time of the error with 1.5M rows, the disk space where the devices are created is as follows:


sdb-fdb-build:[18] $ df -kh /tmp
Filesystem      Size  Used Avail Use% Mounted on
/dev/sda1       128G   93G   35G  73% /
```
So, there is no issue with insufficient disk space.

Initially I thought that this problem is occurring due to running on Nimbus-VMs and with /tmp device. 
But the 2nd assertion is reproducible quite easily on an AWS bare metal instance with 1.25M row count, too. 
(I didn't check the logs there, but suspect that the 2nd assertion occurred during a reboot as a consequence of the 1st assertion in `trunk_verify: pivot kv_bytes incorrect`.
```


### Expected behavior

Expect to be able to run YCSB runs with larger # of row counts, in excess of 1M, 1.5M, 2M and so on.

### Additional context

This problem shows up with SplinterDB integrated as a storage engine behind FDB, but the evidence seems to indicate an existing bug that may occur even for non-FDB workloads.

**Comment by gapisback** [comment]
@ajhconway @rtjohnso -- Just trying to get your attention to this failing assertion.

I an working with a slightly older cut of SplinterDB code base on my build machine `sdb-fdb-build` and am at this SHA b046b03 for trunk.c . 

The offending code fragment is this one:

```
7433 bool
7434 trunk_verify_node(trunk_handle *spl, page_handle *node)
7435 {
[...]
7496       if (trunk_pivot_kv_bytes(spl, node, pivot_no) != kv_bytes) {
7497          platform_error_log("trunk_verify: pivot kv_bytes incorrect\n");
7498          platform_error_log("reported %lu, actual %lu\n",
7499                             trunk_pivot_kv_bytes(spl, node, pivot_no),
7500                             kv_bytes);
```

The l.h.s of the check, `trunk_pivot_kv_bytes()` is simply adding up 2 values:

```
1563 trunk_pivot_kv_bytes(trunk_handle *spl, page_handle *node, uint16 pivot_no)
1564 {
1565    trunk_pivot_data *pdata = trunk_get_pivot_data(spl, node, pivot_no);
1566    return pdata->num_kv_bytes_whole + pdata->num_kv_bytes_bundle;
1567 }
```

How critical do you think this inaccuracy is? 

Could it be a counting / reporting issue, or is it potentially a real-bug in tracking pivots etc.?

In the page dump generated when the assertion trips, I am not sure where the values `pdata->num_kv_bytes_whole`,`pdata->num_kv_bytes_bundle` are printed. So, it's not clear what these two values are (were) when added up to the l.h.s of the check.

I will re-run these tests on AWS bare-metal machine with a release binary just to see if I can YCSB runs w/ larger # of rows working. But I'm afraid that may be masking latent issues.

Comments and suggestions to debug this further are welcome!

**Comment by gapisback** [comment]
Attn: @rtjohnso @ajhconway -- I have been able to repro this bug in the YCSB-C branch with some code changes:

Branch: agurajada/424-repro-Fdb-YCSB

Command used:

```
$ ./ycsbc -threads 10 -db splinterdb -L workloads/load.spec -w recordcount 10000000 -W workloads/workloada.spec -w operationcount 10000000 -p splinterdb.disk_size_gb 40
```
